### PR TITLE
Phase 2: Add metadata, addon packages and improve REST client error handling

### PR DESCRIPTION
**Pull request by alxgsv**
- Add metadata package with full CRUD for file metadata
- Add addon package for executing and polling Addons API (Remove.bg, ClamAV, AWS Rekognition)
- Add `group.Delete()` method for deleting group metadata without deleting files
- Add webhook event constants (`FileStored`, `FileDeleted`, `FileInfoUpdated`, `FileInfected`)
- Return proper errors for unhandled HTTP status codes (>= 400) in the REST client instead of silently ignoring them

New packages:

- `metadata/` - file metadata key-value CRUD, backed by `/files/{uuid}/metadata/` endpoints
- `addon/` - execute addons and poll status via `/addons/{addon}/execute/` and `/addons/{addon}/execute/status/`

REST client fix:

- Previously, HTTP responses with status >= 400 that didn't match a known case (401, 403, 404, 429) fell through to the default branch silently. Now the client attempts to decode the API error body, and if that fails, returns an `unexpectedStatusErr` with the status code.
- Fixed `json.Decode(&resdata)` → `json.Decode(resdata)` (unnecessary double-pointer)

Tests:
  - All new packages have httptest-based unit tests with mock `ucare.Client`
  - Integration test stubs added for metadata and addon operations
  - Unit test for unhandled status code behavior in REST client

## Summary by CodeRabbit

* **New Features**
  * Addons API: execute and poll addon tasks (Remove.bg, ClamAV, Rekognition)
  * File Metadata API: per-file key/value metadata with List/Get/Set/Delete
  * Delete group metadata without removing files
  * New webhook events: file.stored, file.deleted, file.info_updated (file.infected deprecated)

* **Improvements**
  * Config: added UserAgent field for custom agent identification

### CODE REVIEW

File: addon/service.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		statusQuery{RequestID: requestID},
+		&data,
+	)
+	return
+}
+
+type executeBody ExecuteParams
+
+func (p executeBody) EncodeReq(req *http.Request) error {
+	return codec.EncodeReqBody(p, req)
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply addon/service.go:67]: @coderabbitai I checked this path, and the 429 retry should be safe here. These requests are created through the shared request builder, which sets GetBody, and the retry recreates the JSON body by running the encoder again. So this does not depend on EncodeReqBody itself populating GetBody.

File: group/group.go [hunk 2]
```diff
@@ -23,6 +23,22 @@ func (s service) Info(
...
+func (s service) Delete(
+	ctx context.Context,
+	id string,
+) (err error) {
+	err = s.svc.ResourceOp(
+		ctx,
+		http.MethodDelete,
+		fmt.Sprintf(deletePathFormat, id),
+		nil,
+		nil,
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply group/group.go:37]: @coderabbitai good catch. Not specific to this PR tho. I fixed this in `restAPIClient.Do` so `resp.Body` is always closed, including successful requests where `resourceData` is `nil` like this `DELETE` path. I also added a regression test for the `nil` response case.

File: group/group_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	assert "github.com/stretchr/testify/require"
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply group/group_test.go]: @coderabbitai thank you. I replaced the handler assertions with assert to make failures report correctly without relying on fail-fast behavior in that goroutine.

File: metadata/service.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+// Get returns a single metadata value by key
+func (s service) Get(
+	ctx context.Context,
+	fileUUID, key string,
+) (data string, err error) {
+	err = s.svc.ResourceOp(
+		ctx,
+		http.MethodGet,
+		fmt.Sprintf("/files/%s/metadata/%s/", fileUUID, key),
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply metadata/service.go]: @coderabbitai very good find! Fixed

File: metadata/service.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+type stringBody string
+
+func (s stringBody) EncodeReq(req *http.Request) error {
+	raw, err := json.Marshal(string(s))
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewBuffer(raw)
+	req.Body = io.NopCloser(buf)
+	req.ContentLength = int64(buf.Len())
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply metadata/service.go]: @coderabbitai this is already handled. getBodyBuilder wraps EncodeReq so that each GetBody call re-invokes it, creating a fresh bytes.NewBuffer every time.

File: test/addon.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/uploadcare/uploadcare-go/v2/addon"
+	"github.com/uploadcare/uploadcare-go/v2/test/testenv"
+)
+
+func addonClamAVExecute(t *testing.T, r *testenv.Runner) {
+	ctx := context.Background()
+	result, err := r.Addon.Execute(ctx, addon.AddonClamAV, addon.ExecuteParams{
+		Target: r.Artifacts.Files[0].ID,
+		Params: addon.ClamAVParams{},
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply test/addon.go:16]: @coderabbitai it's over-engineering for an integration test helper

File: test/group.go [hunk 7]
```diff
@@ -33,3 +33,9 @@ func groupInfo(t *testing.T, r *testenv.Runner) {
 	assert.Equal(t, r.Artifacts.GroupIDs[0], info.ID)
 }
 
+func groupDelete(t *testing.T, r *testenv.Runner) {
+	ctx := context.Background()
+	err := r.Group.Delete(ctx, r.Artifacts.GroupIDs[0])
+	assert.Equal(t, nil, err)
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply test/group.go:39]: @coderabbitai it's over-engineering for an integration test helper

File: metadata/service_test.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		body, _ := io.ReadAll(resp.Body)
+		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, body)
+	}
+
+	if resdata == nil || reflect.ValueOf(resdata).IsNil() {
+		return nil
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply metadata/service_test.go]: @coderabbitai it's over-engineering for an integration test helper

File: addon/addon.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+type StatusParams struct {
+	// RequestID is the execution request ID returned by Execute
+	RequestID string `json:"request_id"`
+}
+
+// StatusResult holds the response from an addon status check
+type StatusResult struct {
+	// Status is the current execution status
+	Status string `json:"status"`
+
```
Comment (ckcnik) [inline addon/addon.go]: as i see you lost  details field 

```suggestion
Details json.RawMessage `json:"details,omitempty"`
```

File: metadata/validation.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package metadata
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+var (
+	keyPattern = regexp.MustCompile(`^[-_.:A-Za-z0-9]{1,64}$`)
```
Comment (ckcnik) [inline metadata/validation.go]:
```suggestion
	keyPattern = regexp.MustCompile(`^[\w.:-]{1,64}$`)
```
should it be synced with our backend?

File: metadata/validation.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+	// ErrInvalidKey reports metadata keys that do not match the documented format.
+	ErrInvalidKey = errors.New("metadata key must match ^[-_.:A-Za-z0-9]{1,64}$")
+)
```
Comment (ckcnik) [inline metadata/validation.go:23]: in platfrom we have also the follving validation rules:
```
FILE_METADATA_MAX_VALUE_LENGTH = 512
FILE_METADATA_MAX_KEYS_NUMBER = 50
```

Comment (ckcnik) [reply metadata/validation.go:23]: lets declare 

```
const (
	MaxKeyLength   = 64  // FILE_METADATA_MAX_KEY_LENGTH
	MaxValueLength = 512 // FILE_METADATA_MAX_VALUE_LENGTH
	MaxKeysNumber  = 50  // FILE_METADATA_MAX_KEYS_NUMBER
)
```

File: addon/service_test.go [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func TestExecute_WithClamAVParams(t *testing.T) {
+	t.Parallel()
+
+	purge := true
+
+	svc, srv := newTestService(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		assert.NoError(t, err)
```
Comment (ckcnik) [inline addon/service_test.go]:
```suggestion
			require.NoError(t, err)
```

use require everywhere if next checks don't have any sense

File: addon/service_test.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+	return req, nil
+}
+
+func (c *testClient) Do(req *http.Request, resdata interface{}) error {
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
```
Comment (ckcnik) [inline addon/service_test.go]: copy paste, pls create package with these helpers

File: metadata/metadata.go [hunk 14]
```diff
@@ -0,0 +1,6 @@
+// Package metadata holds primitives and logic for the file metadata API.
+//
+// File metadata is a key-value store associated with each file.
+// Keys are strings matching ^[-_.:A-Za-z0-9]{1,64}$ and values are
```
Comment (ckcnik) [inline metadata/metadata.go]: duplicate of https://github.com/uploadcare/uploadcare-go/pull/26/changes#diff-7d45580ec3b063693b7a1f1e7e85669126fa26d8fc91a45f43a5d3c7ba3d04ccR10

File: internal/uctest/client.go [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if data != nil {
+		if err = data.EncodeReq(req); err != nil {
+			return nil, err
+		}
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", fmt.Sprintf(config.AcceptHeaderFormat, ucare.APIv07))
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply internal/uctest/client.go:44]: @coderabbitai this is a REST client, not Upload client

File: internal/uctest/httptest.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func RespondJSON(w http.ResponseWriter, v interface{}) {
+	w.Header().Set("Content-Type", "application/json")
+	_ = json.NewEncoder(w).Encode(v)
+}
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply internal/uctest/httptest.go]: @coderabbitai I don't find it practical, but fixed for consistency

File: ucare/restapi.go [hunk 17]
```diff
@@ -96,80 +96,106 @@ func (c *restAPIClient) NewRequest(
...
 		}
 
 		select {
 		case <-req.Context().Done():
-			return req.Context().Err()
+			return false, req.Context().Err()
 		case <-time.After(time.Duration(retryAfter) * time.Second):
 		}
-		goto try
+		return true, nil
```
[the remark this answers is not in the document]
Comment (alxgsv) [reply ucare/restapi.go:166]: @coderabbitai where did you get `MaxWaitSeconds`?

[the remark this answers is not in the document]
Comment (alxgsv) [reply ucare/restapi.go:166]: @coderabbitai ok fixed absent and negative retry-after paths

File: addon/addon.go [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+type ExecuteParams struct {
+	Target string      `json:"target"`
+	Params interface{} `json:"params,omitempty"`
+}
+
+type ExecuteResult struct {
+	RequestID string `json:"request_id"`
+}
+
+type StatusParams struct {
```
Comment (ckcnik) [inline addon/addon.go:31]: dead code?

File: addon/addon.go [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+const (
+	StatusInProgress = "in_progress"
+	StatusDone       = "done"
+	StatusError      = "error"
+	StatusUnknown    = "unknown"
+)
+
+type ExecuteParams struct {
+	Target string      `json:"target"`
+	Params interface{} `json:"params,omitempty"`
```
Comment (ckcnik) [inline addon/addon.go]:
```suggestion
	Params any `json:"params,omitempty"`
```
interface{}  it's a deprecated style

File: addon/addon.go [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+package addon
+
+import "encoding/json"
+
+const (
+	AddonRemoveBG              = "remove_bg"
+	AddonClamAV                = "uc_clamav_virus_scan"
+	AddonRekognitionLabels     = "aws_rekognition_detect_labels"
+	AddonRekognitionModeration = "aws_rekognition_detect_moderation_labels"
+)
```
Comment (ckcnik) [inline addon/addon.go:13]:
```suggestion
type Name string

const (
	AddonRemoveBG              Name = "remove_bg"
	AddonClamAV                Name = "uc_clamav_virus_scan"
	AddonRekognitionLabels     Name = "aws_rekognition_detect_labels"
	AddonRekognitionModeration Name = "aws_rekognition_detect_moderation_labels"
)
```

much better to use alias - it's more compile-time safety

File: addon/log.go [hunk 21]
```diff
@@ -0,0 +1,18 @@
+package addon
+
+import (
+	"github.com/uploadcare/uploadcare-go/v2/uclog"
+)
+
+var log uclog.Logger
+
+const subsystemTag = "ADON"
```
Comment (ckcnik) [inline addon/log.go:9]:
```suggestion
const subsystemTag = "ADDON"
```
hmm? or it's our convention to use only 4 symbols?

File: group/group.go
Comment (ckcnik) [inline group/group.go:14]:
```suggestion
	id string,
```

File: group/group_test.go [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uploadcare/uploadcare-go/v2/internal/uctest"
+)
+
+func TestDelete(t *testing.T) {
+	t.Parallel()
+
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (ckcnik) [inline group/group_test.go]: pls use uctest.WithHTTPServer( instead of httptest.NewServer

File: metadata/service.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+	return nil
+}
```
Comment (ckcnik) [inline metadata/service.go:114]:
```suggestion
func (s stringBody) EncodeReq(req *http.Request) error {
	return codec.EncodeReqBody(string(s), req)
}
```

File: webhook/service.go [hunk 24]
```diff
@@ -30,7 +30,13 @@ const (
...
 const (
-	EventFileUploaded = "file.uploaded"
+	EventFileUploaded    = "file.uploaded"
+	EventFileStored      = "file.stored"
+	EventFileDeleted     = "file.deleted"
+	EventFileInfoUpdated = "file.info_updated"
+
+	// EventFileInfected is deprecated. Use EventFileInfoUpdated instead.
+	EventFileInfected = "file.infected"
 )
```
Comment (ckcnik) [inline webhook/service.go]:
```
type Event string

const (
	EventFileUploaded    Event = "file.uploaded"
	EventFileStored      Event = "file.stored"
	EventFileDeleted     Event = "file.deleted"
	EventFileInfoUpdated Event = "file.info_updated"

	// EventFileInfected is deprecated. Use EventFileInfoUpdated instead.
	EventFileInfected Event = "file.infected"
)
```

the same like in addons

File: webhook/webhook_test.go [hunk 25]
```diff
@@ -0,0 +1,17 @@
+package webhook
+
+import (
+	"testing"
+
+	assert "github.com/stretchr/testify/require"
+)
+
+func TestEventConstants(t *testing.T) {
```
Comment (ckcnik) [inline webhook/webhook_test.go]: useless test in my opinion, i've added CRUD tests

File: addon/log.go (hunk 21 above)
Comment (alxgsv) [reply addon/log.go:9]: Yes, it's a convention to keep output aligned

**Review by ckcnik — APPROVED** [review]
